vm/src/iterator.rs

```diff
@@ -2,65 +2,24 @@
  * utilities to support iteration.
  */
 
-use crate::builtins::int::{self, PyInt};
-use crate::builtins::iter::PySequenceIterator;
-use crate::builtins::PyBaseExceptionRef;
-use crate::vm::VirtualMachine;
-use crate::{IdProtocol, PyObjectRef, PyResult, PyValue, TypeProtocol};
+use crate::{
+    builtins::{int, PyBaseExceptionRef, PyInt},
+    protocol::PyIter,
+    IdProtocol, PyObjectRef, PyResult, TypeProtocol, VirtualMachine,
+};
 use num_traits::Signed;
 
-/*
- * This helper function is called at multiple places. First, it is called
- * in the vm when a for loop is entered. Next, it is used when the builtin
- * function 'iter' is called.
- */
-pub fn get_iter(vm: &VirtualMachine, iter_target: PyObjectRef) -> PyResult {
-    let getiter = {
-        let cls = iter_target.class();
-        cls.mro_find_map(|x| x.slots.iter.load())
-    };
-    if let Some(getiter) = getiter {
-        let iter = getiter(iter_target, vm)?;
-        let cls = iter.class();
-        let is_iter = cls.iter_mro().any(|x| x.slots.iternext.load().is_some());
-        if is_iter {
-            drop(cls);
-            Ok(iter)
-        } else {
-            Err(vm.new_type_error(format!(
-                "iter() returned non-iterator of type '{}'",
-                cls.name()
-            )))
-        }
-    } else {
-        vm.get_method_or_type_error(iter_target.clone(), "__getitem__", || {
-            format!("'{}' object is not iterable", iter_target.class().name())
-        })?;
-        Ok(PySequenceIterator::new(iter_target)
-            .into_ref(vm)
-            .into_object())
-    }
-}
-
-pub fn call_next(vm: &VirtualMachine, iter_obj: &PyObjectRef) -> PyResult {
-    let iternext = {
-        let cls = iter_obj.class();
-        cls.mro_find_map(|x| x.slots.iternext.load())
-            .ok_or_else(|| {
-                vm.new_type_error(format!("'{}' object is not an iterator", cls.name()))
-            })?
-    };
-    iternext(iter_obj, vm)
-}
-
 /*
  * Helper function to retrieve the next object (or none) from an iterator.
  */
-pub fn get_next_object(
+pub fn get_next_object<T>(
```

Comment:
Is my understanding correct that the remaining functions in `iterator` will (eventually) not be part of the public API? If so, is there a reason keeping them here instead of having all iterator related functions in `protoco/iter.rs`?

Reply:
I think these functions will be part of public API. The major reason I didn't do it in this PR is to confirm the design by review before going further. I will do more work in future with new PRs.

In CPython, The public API `PyIter_Next` returns (if we can say so) `PyResult<Option<PyObjectRef>>` as its primitive. But in RustPython, `PyIter::next` returns `PyResult<PyObejctRef>` as its primitive. So there will be more decision works not just about `PyIter` itself.

This is related to #3124